packages/hardhat/test/e2e/periphery/controller-helper.ts

```diff
@@ -444,6 +445,127 @@ describe("ControllerHelper: mainnet fork", function () {
+        sqrtPriceLimitX96: 0
+      }    
+
+      const ownerSqueethBalanceBeforeTrade2 = await wSqueeth.balanceOf(owner.address)
+      await wSqueeth.connect(owner).approve(uniswapRouter.address, constants.MaxUint256)
+      await uniswapRouter.connect(owner).exactInputSingle(swapParamSell)
+      const ownerSqueethBalanceAfterTrade2 = await wSqueeth.balanceOf(owner.address)
+      expect(ownerSqueethBalanceBeforeTrade2.sub(ownerSqueethBalanceAfterTrade2).eq(wSqueethToSell)).to.be.true
+
+      const vaultId = (await shortSqueeth.nextId()).sub(1);
+      const uniTokenId =  (await controller.vaults(vaultId)).NftCollateralId;
+      const vaultBefore = await controller.vaults(vaultId); 
+      const ethPrice = await oracle.getTwap(ethUsdcPool.address, weth.address, usdc.address, 420, true)
+      const scaledEthPrice = ethPrice.div(10000)
+      const debtInEth = vaultBefore.shortAmount.mul(scaledEthPrice).div(one)
+      const collateralToFlashloan = debtInEth.mul(3).div(2).add(ethers.utils.parseUnits('0.01'))
+      const positionBefore = await (positionManager as INonfungiblePositionManager).positions(uniTokenId);
+
+      const amount0Max = BigNumber.from(2).mul(BigNumber.from(10).pow(18)).sub(1)
+      const amount1Max = BigNumber.from(2).mul(BigNumber.from(10).pow(18)).sub(1)
+
+      const abiCoder = new ethers.utils.AbiCoder
+      const params = [
+      {
+          rebalanceVaultNftType: BigNumber.from(6),
+          // data: ethers.utils.hexlify(abiCoder.encode(["uint256"], ["1"])) as BytesLike
+          data: abiCoder.encode(["uint256", "uint128", "uint128"], [uniTokenId, amount0Max, amount1Max])
+        },
+        {
+          rebalanceVaultNftType: BigNumber.from(7),
+          // data: ethers.utils.hexlify(abiCoder.encode(["uint256"], ["1"])) as BytesLike
+          data: abiCoder.encode(["uint256"], [uniTokenId])
+        }
+      ]
+
+      await controller.connect(depositor).updateOperator(vaultId, controllerHelper.address);
+
+      const depositorSqueethBalanceBefore = await wSqueeth.balanceOf(depositor.address)
+      const depositorEthBalanceBefore = await ethers.provider.getBalance(depositor.address)
+      await controllerHelper.connect(depositor).rebalanceVaultNft(vaultId, collateralToFlashloan, params);
+      const depositorSqueethBalanceAfter = await wSqueeth.balanceOf(depositor.address)
+      const depositorEthBalanceAfter = await ethers.provider.getBalance(depositor.address)
+
+      const positionAfter = await (positionManager as INonfungiblePositionManager).positions(uniTokenId);
+      const vaultAfter = await controller.vaults(vaultId); 
+
+      expect(positionAfter.tickLower === -887220).to.be.true
+      expect(positionAfter.tickUpper === 887220).to.be.true
+      expect(positionAfter.liquidity.eq(positionBefore.liquidity)).to.be.true
+      expect(vaultAfter.NftCollateralId==vaultBefore.NftCollateralId).to.be.true
+      expect(depositorSqueethBalanceAfter.gt(depositorSqueethBalanceBefore)).to.be.true
```

Comment:
Tried to the difference using accumulated fees from positions like this:

      const tokensOwed0 = positionBefore.feeGrowthInside0LastX128.mul(positionBefore.liquidity).div(BigNumber.from(2).pow(128));
      const tokensOwed1 = positionBefore.feeGrowthInside1LastX128.mul(positionBefore.liquidity).div(BigNumber.from(2).pow(128));

But doesn't work for some reason. What am I doing wrong here? Is there another way to get it?

Reply:
I think you need to update state first.  feeGrowthInside0LastX128 is not updated for all positions on each trade, but rather updated from a global variable.

You can trace it from here: 
https://github.com/Uniswap/v3-periphery/blob/main/contracts/NonfungiblePositionManager.sol#L330
https://github.com/Uniswap/v3-core/blob/main/contracts/UniswapV3Pool.sol#L440